sockeye/inference.py

```diff
@@ -1043,25 +1047,58 @@ def _beam_search(self,
+            scores_accumulated = mx.nd.where(newly_finished, scores_accumulated / self.length_penalty(lengths), scores_accumulated)
+
+            # (5) Prune out low-probability hypotheses
+            for sentno in range(self.batch_size):
+
+                # Set the active beam size to full. It will get reduced if items are filtered.
+                active_beam_size[sentno] = self.beam_size
+
+                rows = slice(sentno * self.beam_size, (sentno + 1) * self.beam_size)
+                if self.beam_prune > 0.0:
+                    if mx.nd.sum(finished[rows]) > 0:
+                        best_score = mx.nd.min(mx.nd.where(finished[rows].expand_dims(axis=1), scores_accumulated[rows], inf_array))
+
+                        to_remove = mx.nd.cast(scores_accumulated[rows,0] - best_score > self.beam_prune, dtype='int32')
+                        scores_accumulated[rows] = mx.nd.where(to_remove, inf_array, scores_accumulated[rows])
+                        best_word_indices[rows] = mx.nd.where(to_remove, zeros_array, best_word_indices[rows])
+
+                        num_removing = int(mx.nd.sum(to_remove).asscalar())
+                        active_beam_size[sentno] = self.beam_size - num_removing
+
+                        # mark removed ones as finished so they won't block early exiting
+                        finished[rows] = finished[rows] + to_remove
+
+                # Resort the hypotheses
+                sorted_indices = mx.nd.argsort(scores_accumulated[rows,0])
+                best_hyp_indices[rows] = mx.nd.take(best_hyp_indices[rows], sorted_indices)
+                best_word_indices[rows] = mx.nd.take(best_word_indices[rows], sorted_indices)
+                scores_accumulated[rows] = mx.nd.take(scores_accumulated[rows,0], sorted_indices.expand_dims(axis=1))
+
+            # (6) get hypotheses and their properties for beam_size winning hypotheses (ascending)
             sequences = mx.nd.take(sequences, best_hyp_indices)
             lengths = mx.nd.take(lengths, best_hyp_indices)
             finished = mx.nd.take(finished, best_hyp_indices)
             attention_scores = mx.nd.take(attention_scores, best_hyp_indices)
             attentions = mx.nd.take(attentions, best_hyp_indices)
 
-            # (5) update best hypotheses, their attention lists and lengths (only for non-finished hyps)
+            # (7) update best hypotheses, their attention lists and lengths (only for non-finished hyps)
             # pylint: disable=unsupported-assignment-operation
             sequences[:, t] = best_word_indices
             attentions[:, t, :] = attention_scores
             lengths += mx.nd.cast(1 - mx.nd.expand_dims(finished, axis=1), dtype='float32')
 
-            # (6) determine which hypotheses in the beam are now finished
             finished = ((best_word_indices == C.PAD_ID) + (best_word_indices == self.vocab_target[C.EOS_SYMBOL]))
```

Comment:
Does this logic still hold with your change? (you created a new finished array intermediately that may differ).